kvserver: refactor replicate queue to enable allocator code reuse

AlexTalks opened the pull request:
This change refactors parts of the replicate queue's `PlanOneChange(..)`
and `addOrRemove{Non}Voters(..)` functions to reusable helper functions
that simplify usage of the allocator and deduplicate repeated code
paths. The change also adds convenience methods to the `AllocatorAction`
enum, to move certain determinations (such as if a computed allocator
action is a remove or a replace) closer to the allocator type it is
based on. These changes further enable the ability to use the allocator
outside of the replicate queue, and enable the ability for some of this
logic to move into the allocator itself in the future.

Depends on #91941.

Epic: none

Release note: None

AlexTalks committed:
kvserver: refactor replicate queue allocator usage logic

This change refactors parts of the replicate queue's `PlanOneChange(..)`
and `addOrRemove{Non}Voters(..)` functions to reusable helper functions
that simplify usage of the allocator and deduplicate repeated code
paths. The change also adds convenience methods to the `AllocatorAction`
enum, to move certain determinations (such as if a computed allocator
action is a remove or a replace) closer to the allocator type it is
based on. These changes move more of the logic needed to use the
allocator into the `allocatorimpl` package itself, enabling usage of the
allocator outside of the replicate queue.

Part of #91571.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94114)
<!-- Reviewable:end -->

kvoli commented on pkg/kv/kvserver/replicate_queue.go:
```diff
@@ -1190,9 +1150,109 @@ func (rq *replicateQueue) PlanOneChange(
+// In case of an add action, no replicas are removed and -1 is returned, and if no
+// candidates for replacement can be found during a replace action, the returned
+// nothingToDo flag will be set to true.
+// TODO(sarkesian): If possible, move this logic into the allocator.
```
We should certainly  try  and  move this into the `allocatorimpl` pkg.

AlexTalks committed:
kvserver: refactor upreplication fragile quorum check into allocator

While previously we checked that we can avoid a fragile quorum during
the addition of new voters in the replicate queue, this change moves the
check logic into the allocator code itself, allowing it to be reused by
other users of the allocator. This enables us to perform this check when
evaluating decommission viability, or anywhere else that uses the
allocator for new voter replicas.

Part of #91570.

Release note: None

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NJr2YhrJXNt8pJFRjjK-r1-1178:-NJvqTaT7ULS_ZfiBDPS:b-fajfam) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdce47e413dbc66d1ca05d1d6c61f083e1a4afa1/pkg/kv/kvserver/replicate_queue.go#L1178)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

We should certainly  try  and  move this into the `allocatorimpl` pkg. 
</blockquote></details>

Done.

Also moved the "avoid fragile quorum" check into the allocator as well.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
Looks good. There is one test failing that I would look into:

TestPromoteNonVoterInAddVoter

There is fragile logic which decides to promote a non-voter on removal.

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-:-NKJBRVtDpR6Jy4aJ3lK:bepuy2c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NKJAtJA7psEVxClXRmk:bu509yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
> ```Go
> ) (op AllocationOp, _ error) {
> 	effects := effectBuilder{}
> 	conf := repl.SpanConfig()
> ```

This should be passed in to be consistent with the other commit that makes desc/conf consistent. 


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL2kPI0CgyX22Hr6CHQ:b-x4h0cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This should be passed in to be consistent with the other commit that makes desc/conf consistent. 
</blockquote></details>

Some work will be required to rebase that change on top of all the other changes to the replicate queue/allocator, so I'll wait until these are merged to do that.


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-:-NL8eq7PBTBCJ9Oapx2t:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NJr2YhrJXNt8pJFRjjK-r1-1178:-NL8eSNe6O84zkOvh7u1:btojbfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdce47e413dbc66d1ca05d1d6c61f083e1a4afa1/pkg/kv/kvserver/replicate_queue.go#L1178)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Done.

Also moved the "avoid fragile quorum" check into the allocator as well.
</blockquote></details>

Great!
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL8e4R42vG6BdAGk1DR:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Some work will be required to rebase that change on top of all the other changes to the replicate queue/allocator, so I'll wait until these are merged to do that.
</blockquote></details>

ack
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1038 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8elzX1wwUZeqxdADY:-NL8elzX1wwUZeqxdADZ:btni5mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/replicate_queue.go#L1038)):*
> ```Go
> 				action,
> 				voterReplicas, nonVoterReplicas,
> 				liveVoterReplicas, deadVoterReplicas,
> ```

Just curious - was this the source of the test failures previously?
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8e7md8nrTUqL_jyBO:-NL8e7md8nrTUqL_jyBP:b-tpfh90) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L136)):*
> ```Go
> )
> 
> func (a AllocatorAction) Add() bool {
> ```

nit: does the linter enforce export function commenting? These are pretty self explanatory, just wondering why that got removed.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL92hg1AzUUJzBtSWLv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

ack
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1038 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8elzX1wwUZeqxdADY:-NL92XZV9U38z58sGrcV:b-5fhm6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/replicate_queue.go#L1038)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Just curious - was this the source of the test failures previously?
</blockquote></details>

One of them, unfortunately.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8e7md8nrTUqL_jyBO:-NL92kdZ9LDPB-k57DeS:b-ng64w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L136)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: does the linter enforce export function commenting? These are pretty self explanatory, just wondering why that got removed.
</blockquote></details>

I don't think so, can add if needed.


<!-- Sent from Reviewable.io -->

AlexTalks commented:
bors r+
